formal-models/dbft/dbft.tla

```diff
@@ -0,0 +1,366 @@
+  /\ Cardinality(RMFault \cup RMDead) <= F
+  /\ MaxView \in Nat
+  /\ MaxView <= 2
+
+\* RMStates is a set of records where each record holds the node state and
+\* the node current view.
+RMStates == [
+              type: {"initialized", "prepareSent", "commitSent", "cv", "blockAccepted", "bad", "dead"},
+              view : Nat
+            ]
+
+\* Messages is a set of records where each record holds the message type,
+\* the message sender and sender's view by the moment when message was sent.
+Messages == [type : {"PrepareRequest", "PrepareResponse", "Commit", "ChangeView"}, rm : RM, view : Nat]
+
+\* -------------- Useful operators --------------
+
+\* IsPrimary is an operator defining whether provided node r is primary
+\* for the current round from the r's point of view. It is a mapping
+\* from RM to the set of {TRUE, FALSE}.
+IsPrimary(r) == rmState[r].view % N = r
+
+\* GetPrimary is an operator defining mapping from round index to the RM that
+\* is primary in this round.
+GetPrimary(view) == CHOOSE r \in RM : view % N = r
+
+\* GetNewView returns new view number based on the previous node view value.
+\* Current specifications only allows to increment view.
+GetNewView(oldView) == oldView + 1
+
+\* CountCommitted returns the number of nodes that have sent the Commit message
+\* in the current round (as the node r sees it).
+CountCommitted(r) == Cardinality({rm \in RM : Cardinality({msg \in msgs : msg.rm = rm /\ msg.type = "Commit" /\ msg.view = rmState[r].view}) /= 0})
+
+\* MoreThanFNodesCommitted returns whether more than F nodes have been committed
+\* in the current round (as the node r sees it).
+MoreThanFNodesCommitted(r) == CountCommitted(r) > F
+
+\* PrepareRequestSentOrReceived denotes whether there's a PrepareRequest
+\* message received from the current round's speaker (as the node r sees it).
+PrepareRequestSentOrReceived(r) == [type |-> "PrepareRequest", rm |-> GetPrimary(rmState[r].view), view |-> rmState[r].view] \in msgs
+
+\* -------------- Safety temporal formula --------------
+
+\* Init is the initial predicate initializing values at the start of every
+\* behaviour.
+Init ==
+  /\ rmState = [r \in RM |-> [type |-> "initialized", view |-> 0]]
+  /\ msgs = {}
+
+\* RMSendPrepareRequest describes the primary node r broadcasting PrepareRequest.
+RMSendPrepareRequest(r) ==
+  /\ rmState[r].type = "initialized"
+  /\ IsPrimary(r)
+  /\ rmState' = [rmState EXCEPT ![r].type = "prepareSent"]
+  /\ msgs' = msgs \cup {[type |-> "PrepareRequest", rm |-> r, view |-> rmState[r].view]}
+  /\ UNCHANGED <<>>
+
+\* RMSendPrepareResponse describes non-primary node r receiving PrepareRequest from
+\* the primary node of the current round (view) and broadcasting PrepareResponse.
+\* This step assumes that PrepareRequest always contains valid transactions and
+\* signatures.
+RMSendPrepareResponse(r) ==
+  /\ \/ rmState[r].type = "initialized"
+     \/ /\ rmState[r].type = "cv"
+        /\ MoreThanFNodesCommitted(r)
```

Comment:
This transition is one of most unobvious ones, probably deserves a comment (relates to NotAcceptingPayloadsDueToViewChanging logic).

Reply:
Fixed. Do we need to write a couple of words about it in the README? I've added a comment in the model itself.